### ISSUE TITLE
Better n-up support and other POD support for save_sheet

### ISSUE THREAD

**Issue by andymeneely** [andymeneely/squib#126]
Initial description from BGG user azarkon:

> simple: add/modify save_sheet to have useful defaults: 9-up print and play(no bleed), 8-up print and play (full bleed), POD compliant formats (gamecrafter, PNP productions, and Drive-through cards, etc)

We can also make DSL methods for specific manufacturers too.

**Comment by andjd** [comment]
I'd like to make bleed/trim an instance variable of the Deck class with a default value of 0.  I feel that it's a detail that's conceptually immutable like deck size and dimensions, and this would allow me to write the save sheet method so it wouldn't need a trim argument. We could then change all the render methods to take a :bleed argument that would equal :full, :none, or :safe too.  

thoughts?

**Comment by andymeneely** [comment]
The default `trim` option to `save_sheet` is already 0. 

(Instance variable isn't the way to accomplish this anyway)

I don't like full/none/safe as being an options here. That would make the Squib code too dependent on users' needs and I don't want to assume. Can just as easily be defined in a layout file. Perhaps we can include built-in layout files for popular templates. When it comes to options to methods, I'd rather have a minimal set of options, and `trim`, `gap`, and `margin` seem to be enough.

What I do currently is something like this:

```yml
# layout.yml
cut_line:
  x: 37.5
  y: 37.5
  width: 750
  height: 1050
  radius: 0
  stroke_width: 3.0
  stroke_color: black
  trim: 37.5
outline:
  extends: cut_line
  radius: 25
  stroke_width: 6.0
```

Then my code still has the word "cut_line" to make it more readable:
```ruby
rect layout: 'cut_line'
```

Now currently, `save_sheet` doesn't take `trim` from a layout file for some silly reason. I think that's what we really want. That would mean we could do:

```ruby
save_sheet layout: 'cut_line'
```

**Comment by andymeneely** [comment]
As I'm working on using `save_sheet` for the PNP Productions template I'm realizing the margins are different for different sides.  There are probably a number of fixed-document type of issues I'd run into. Trying to decide if this is really a job for save_sheet or if it needs its own method. And then maybe there's some refactoring to be done from there. 

By the way, I'm working on this in the "publishers" branch if you want to take a look.

**Comment by andjd** [comment]
I would have used margin_[top|bottom|right|left].  It's less ambiguous nomenclature and consistent with CSS terminology.

**Comment by andjd** [comment]
What do you think about letting users select a page size, and then packing as many cards on the page as can fit.  for example:

`save_sheet page_size: :letter, gap: '.125in', margin: 150`

**Comment by andymeneely** [comment]
Wait... east/west is ambiguous and left/right is not?!?!? "If you want to get to my house, I'm on the left side of town".

But seriously... yeah sure I'll change that. 

Yeah the page size thing is something I've been meaning to do. The Cairo gem has support for that built-in, but I haven't gotten around to deciphering it. https://github.com/rcairo/rcairo/blob/bb2c21f932d2693f7305bdda1e28973ea08270cb/lib/cairo/papers.rb 

I'm gonna create a separate issue for that - it's easy enough to do in isolation. #135

**Comment by amyworrall** [comment]
Will this new support work with the irregular card spacing of Print And Play Productions? The margins vary depending where the card is in the sequence.

<img width="1029" alt="screen shot 2016-04-05 at 00 10 59" src="https://cloud.githubusercontent.com/assets/606458/14266808/fb27b00a-fac2-11e5-8bd0-c059acbaf0cf.png">

**Comment by andymeneely** [comment]
No - right now the code assumes regularly spaced. That's why I'm thinking it's the wrong approach. Maybe instead we have it based on a yaml specification that gets bundled with Squib. Make it an array of hashes that all have x-y coordinates or something. I haven't take the time to figure out the details, though.

**Comment by amyworrall** [comment]
Yeah, the array would certainly be flexible enough. 

It would be awesome if there were prebuilt templates for all of their card sizes. Happy to help calculate the values if we do go for that approach!

**Comment by andymeneely** [comment]
Actually... could you help with that @amyworrall? Let's get this thing done because I've been super busy (actually designing a few games, plus, you know, life). Here's what we need:

  * Which publisher and type we're talking about, e.g. pnp productions
  * x-y coordinates for the upper-left of each card. Do it in pixels.
  * Intended card size so we can check and/or auto-trim and/or scale the cards to size.
  * Order in which we lay out the cards.
  * Card backs? Still figuring out how we want to handle this with #93 but the data would be nice.

Just post the coordinates here in whatever format you like. I'll credit you in the changelog when it comes together.

**Comment by amyworrall** [comment]
For measurements in pixels, what dpi are we talking? I normally work in points or millimetres for print stuff, since I'm usually sending off PDFs…

**Comment by andymeneely** [comment]
300 ppi is the standard, although let's just default to whatever the printer people are expecting. So the pnp productions one is just in PNG - just use the pixels for that.

**Comment by andymeneely** [comment]
Actually pnp productions is probably a bad example because it looks like they don't have a standard template. They just ask for crop marks so they know where to cut. http://www.printplaygames.com/product/cards. Crop marks are a different feature #123.

**Comment by amyworrall** [comment]
They have standard templates for their non-custom sizes (e.g. "Mini Cards" — that's the one I screenshotted above.)

**Comment by andymeneely** [comment]
Oh right - this one. http://www.printplaygames.com/product/2-5-by-3-5-cards-18-card-sheet

Yeah let's start with those.

**Comment by andymeneely** [comment]
Some more companies we should consider supporting:
  * superiorpod.com http://support.superiorpod.com/customer/portal/articles/726246-spod-printing-templates
  * drive thru cards https://onebookshelfpublisherservice.zendesk.com/entries/94285157-Preparing-Your-Cards

These don't need explicit support for save_sheet or save_pdf because they support individual files
  * thegamecrafter.com
  * printerstudio.com

But in #136 we're talking about having some presets there.

**Comment by amyworrall** [comment]
OK, not sure if this is info in the right format, but let's try it.

**Publisher**: Print & Play Productions
**Card type**: Mini Cards
**Card size**: 1.75 inch x 2.5 inch
**Document size**: 12 inch x 18 inch (contains 36 cards)
**Order**: start top left, lay out left-to-right then top-to-bottom

**Card positions**: (in points, with 72 points == 1 inch):
_NB, the positions are described as {{left, top}, {width, height}}, in points. The positions do not include the bleed area. 9 points (1/8 inch) of bleed should be added to the cards on the three sides specified here._

> 0: {{36, 63}, {126, 180}}, bleed top,left,bottom
> 1: {{162, 63}, {126, 180}}, bleed top,right,bottom
> 2: {{306, 63}, {126, 180}}, bleed top,left,bottom
> 3: {{432, 63}, {126, 180}}, bleed top,right,bottom
> 4: {{576, 63}, {126, 180}}, bleed top,left,bottom
> 5: {{702, 63}, {126, 180}}, bleed top,right,bottom
> 6: {{36, 261}, {126, 180}}, bleed top,left,bottom
> 7: {{162, 261}, {126, 180}}, bleed top,right,bottom
> 8: {{306, 261}, {126, 180}}, bleed top,left,bottom
> 9: {{432, 261}, {126, 180}}, bleed top,right,bottom
> 10: {{576, 261}, {126, 180}}, bleed top,left,bottom
> 11: {{702, 261}, {126, 180}}, bleed top,right,bottom
> 12: {{36, 459}, {126, 180}}, bleed top,left,bottom
> 13: {{162, 459}, {126, 180}}, bleed top,right,bottom
> 14: {{306, 459}, {126, 180}}, bleed top,left,bottom
> 15: {{432, 459}, {126, 180}}, bleed top,right,bottom
> 16: {{576, 459}, {126, 180}}, bleed top,left,bottom
> 17: {{702, 459}, {126, 180}}, bleed top,right,bottom
> 18: {{36, 657}, {126, 180}}, bleed top,left,bottom
> 19: {{162, 657}, {126, 180}}, bleed top,right,bottom
> 20: {{306, 657}, {126, 180}}, bleed top,left,bottom
> 21: {{432, 657}, {126, 180}}, bleed top,right,bottom
> 22: {{576, 657}, {126, 180}}, bleed top,left,bottom
> 23: {{702, 657}, {126, 180}}, bleed top,right,bottom
> 24: {{36, 855}, {126, 180}}, bleed top,left,bottom
> 25: {{162, 855}, {126, 180}}, bleed top,right,bottom
> 26: {{306, 855}, {126, 180}}, bleed top,left,bottom
> 27: {{432, 855}, {126, 180}}, bleed top,right,bottom
> 28: {{576, 855}, {126, 180}}, bleed top,left,bottom
> 29: {{702, 855}, {126, 180}}, bleed top,right,bottom
> 30: {{36, 1053}, {126, 180}}, bleed top,left,bottom
> 31: {{162, 1053}, {126, 180}}, bleed top,right,bottom
> 32: {{306, 1053}, {126, 180}}, bleed top,left,bottom
> 33: {{432, 1053}, {126, 180}}, bleed top,right,bottom
> 34: {{576, 1053}, {126, 180}}, bleed top,left,bottom
> 35: {{702, 1053}, {126, 180}}, bleed top,right,bottom

If it would be easier to see the algorithm for describing the card positions, here it is. (In Objective-C, since that was easiest for me to easily render a PDF to check things.)

```
  CGFloat cardwidth = 1.75;
  CGFloat cardHeight = 2.5;
  CGFloat cardPairGap = 0.25;
  CGFloat leftMargin = 0.5;
  CGFloat topMargin = 0.875;

  CGFloat pointsPerInch = 72.0;
  CGFloat pageWidth = 12;
  CGFloat pageHeight = 18;

  for (int i=0; i<36; i++) {
    CGFloat xOrdinal = i%6;
    CGFloat yOrdinal = i/6;

    CGFloat xPos = leftMargin + xOrdinal * cardwidth + (CGFloat)((int)xOrdinal/2) * cardPairGap;
    CGFloat yPos = topMargin + yOrdinal * cardHeight + (CGFloat)yOrdinal * cardPairGap;

    CGRect cardRect = CGRectMake(xPos * pointsPerInch,
                                 yPos * pointsPerInch,
                                 cardwidth * pointsPerInch,
                                 cardHeight * pointsPerInch);

    NSLog(@"%i: %@, bleed %@", i, NSStringFromCGRect(cardRect), i%2==0 ? @"top,left,bottom" : @"top,right,bottom");
  }

```

**Comment by andymeneely** [comment]
Great! I think this is all the info we need. Just to save me from confusion - can you include a link in each of these reports to the original template? We can eventually put that into our docs too. 

Yeah and even if the units are points (1/72) or pixels or whatever - we can always just translate them. The way Cairo works is you just scale the entire coordinate system to whatever unit conversion you need so our templates could even support a ppi or a dpi option.

By the way, are you amorya on BGG? I would like to give you some GG for your time.

**Comment by andymeneely** [comment]
Here's my current thinking for what the DSL method could look like. 

```ruby
save_pnp_productions template: :mini, trim: '0.125in' to: 'pnp.pdf', format: :pdf
save_pnp_productions to: 'pnp.pdf', format: :png
```
PDF might be better than PNG because it supports vector - depends on how good/bad PNP Productions' printers are compared to Cairo's rasterization. (See my discussion on that [here](http://squib.readthedocs.org/en/latest/backends.html))

And then, to connect this functionality to what #136 was going for, the `preset:` option for `Squib::Deck.new` would automatically set the default for `preset:` and `trim:` using the idea behind #131. So it could be something like this:

```ruby
Squib::Deck.new(preset: :pnp_productions_mini) do
  # do all your stuff
  save_pnp_productions # custom defaults already set, but you can customize here if you want
end
```

That would be equivalent to:
```ruby
Squib::Deck.new do
  default trim: '0.125in'
  default template: :mini

  # do all your stuff

  save_pnp_productions # custom defaults already set, but you can customize here if you want
end
```

Thoughts?

**Comment by amyworrall** [comment]
> PDF might be better than PNG because it supports vector - depends on how good/bad PNP Productions' printers are compared to Cairo's rasterization.

I've only ever sent them vector PDFs. In the past I've used InDesign and Illustrator to create them. I've been quite happy with the quality so far.

> can you include a link in each of these reports to the original template? 

I have updated my post. I've also updated it with some info for how to do card backs.

> By the way, are you amorya on BGG? I would like to give you some GG for your time.

Yep, that's me :)

**Comment by andymeneely** [comment]
Fun coincidence: I ran into a situation in one of my own prototypes where I needed to print to a custom Avery sticker sheet. I wrote up a basic layout sheet thing that takes a specification from Yaml and saves it to a sheet. It's missing some polish to be in Squib at the moment, but it's a start. Feel free to borrow the code: 

https://github.com/andymeneely/project-timber-wolf/blob/master/src/save_custom_sheet.rb
https://github.com/andymeneely/project-timber-wolf/blob/master/data/avery_5408.yml
